elliptic-curve/src/scalar.rs

```diff
@@ -0,0 +1,48 @@
+//! Scalar types
+
+use core::borrow::Borrow;
+use subtle::{ConstantTimeEq, CtOption};
+
+#[cfg(feature = "zeroize")]
+use zeroize::Zeroize;
+
+/// Non-zero scalar type.
+///
+/// This type ensures that its value is not zero, ala `core::num::NonZero*`.
+/// To do this, the generic `S` type must impl both `Default` and
+/// `ConstantTimeEq`, with the requirement that `S::default()` returns 0.
+///
+/// In the context of ECC, it's useful for ensuring that scalar multiplication
+/// cannot result in the point at infinity.
+pub struct NonZeroScalar<S: Default + ConstantTimeEq>(S);
+
+impl<S> NonZeroScalar<S>
+where
+    S: Default + ConstantTimeEq,
+{
+    /// Create a [`NonZeroScalar`] from a scalar, performing a constant-time
+    /// check that it's non-zero.
+    pub fn new(scalar: S) -> CtOption<Self> {
+        let is_some = scalar.ct_eq(&S::default());
```

Comment:
It's a bit weird using `default()` and not `zero()`/`is_zero()`, but I guess it's the best that can be done at the moment. Should probably be refactored when there's a fully-functional `Scalar` trait (or a collection of traits) is available.

Reply:
Yeah, it's the only extant trait to define this in terms of right now.

However, I wouldn't say it's "weird" per say, at least for usage of `Default` in Rust: zero is the `Default` value for all numerical types.

Reply:
I suppose it's ok for simple types, but this approach may fail if `Scalar` ever implements lazy reduction arithmetic similar to that in `FieldElement` (and it will fail for the current `AffinePoint`/`ProjectivePoint`), because in that case you could have something that is zero/identity and yet isn't `default()`.